src/modules/roc_fec/reader.cpp

```diff
@@ -89,61 +90,45 @@ packet::PacketPtr Reader::read_() {
     return get_next_packet_();
 }
 
-packet::PacketPtr Reader::read_first_packet_() {
+packet::PacketPtr Reader::get_first_packet_() {
     packet::PacketPtr pp = source_queue_.head();
-    if (pp) {
-        if (!has_source_) {
-            source_ = pp->rtp()->source;
-            has_source_ = true;
-        }
-
-        cur_sbn_ = pp->fec()->source_block_number;
-        if (!update_size_(pp->fec()->source_block_length)) {
-            return NULL;
-        }
-
-        drop_repair_packets_from_prev_blocks_();
+    if (!pp) {
+        return NULL;
     }
 
-    return pp;
-}
-
-bool Reader::update_size_(size_t new_sblen) {
-    const size_t cur_sblen = source_block_.size();
-    if (cur_sblen == new_sblen) {
-        return true;
-    }
+    packet::FEC& fec = *pp->fec();
+    packet::RTP& rtp = *pp->rtp();
 
-    if (next_packet_ != 0) {
-        roc_log(LogDebug,
-                "fec reader: can't update source block size in the middle"
-                " of the current block, shutting down:"
-                " esi=%lu cur_sbl=%lu new_sbl=%lu",
-                (unsigned long)next_packet_, (unsigned long)cur_sblen,
-                (unsigned long)new_sblen);
-        return (alive_ = false);
+    cur_sbn_ = fec.source_block_number;
+    if (!update_block_size_(fec.source_block_length)) {
+        roc_log(LogTrace,
+                "fec reader: dropping first source packet: can't update block size: "
+                "esi=%lu sblen=%lu blen=%lu",
+                (unsigned long)fec.encoding_symbol_id,
+                (unsigned long)fec.source_block_length, (unsigned long)fec.block_length);
+        return NULL;
     }
 
-    if (!source_block_.resize(new_sblen)) {
-        roc_log(LogDebug,
-                "fec reader: can't resize source block, shutting down:"
-                " cur_sbl=%lu new_sbl=%lu",
-                (unsigned long)cur_sblen, (unsigned long)new_sblen);
-        return (alive_ = false);
+    if (!has_source_) {
+        source_ = rtp.source;
+        has_source_ = true;
     }
 
-    roc_log(LogDebug, "fec reader: update sblen, cur_sbl=%lu new_sbl=%lu",
-            (unsigned long)cur_sblen, (unsigned long)new_sblen);
+    drop_repair_packets_from_prev_blocks_();
 
-    return true;
+    return pp;
 }
 
 packet::PacketPtr Reader::get_next_packet_() {
-    update_packets_();
+    update_block_();
 
     packet::PacketPtr pp = source_block_[next_packet_];
 
     do {
+        if (!alive_) {
+            break;
+        }
+
         if (!pp) {
             try_repair_();
```

Comment:
Discussed to leave it as is for now, because 1) it's not a big problem, we just waste one more cycle in case of a (rare) error, and it is handled correctly in toplevel read(); 2) there are plenty other pieces in reader where the same approach is used and fixing all of them will complicate the code.